File: python3/bron_kerbosch_pivot.py
# coding: utf-8

from graph import UndirectedGraph, Vertex
from consumer import CliqueConsumer

from typing import Callable, List, Set

PivotChoice = Callable[[UndirectedGraph, Set[Vertex]], Vertex]


def visit(
    graph: UndirectedGraph,
    consumer: CliqueConsumer,
    pivot_choice_X: bool,
    candidates: Set[Vertex],
    excluded: Set[Vertex],
    clique: List[Vertex],
) -> None:
    assert all(graph.has_degree(v) for v in candidates)
    assert all(graph.has_degree(v) for v in excluded)
    assert candidates.isdisjoint(excluded)
    assert len(candidates) >= 1
    if len(candidates) == 1:
        # Same logic as below, stripped down for this common case
        for v in candidates:
            neighbours = graph.adjacencies[v]
            assert neighbours
            if excluded.isdisjoint(neighbours):
                consumer.accept(clique + [v])
        return

    # Quickly handle locally unconnected candidates while finding pivot
    remaining_candidates = []
    seen_local_degree = 0
    for v in candidates:
        neighbours = graph.adjacencies[v]
        local_degree = len(candidates.intersection(neighbours))
        if local_degree == 0:
            # Same logic as below, stripped down
            if neighbours.isdisjoint(excluded):
                consumer.accept(clique + [v])
        else:
            if seen_local_degree < local_degree:
                seen_local_degree = local_degree
                pivot = v
            remaining_candidates.append(v)
    if seen_local_degree == 0:
        return
    if pivot_choice_X:
        for v in excluded:
            neighbours = graph.adjacencies[v]
            local_degree = len(candidates.intersection(neighbours))
            if seen_local_degree < local_degree:
                seen_local_degree = local_degree
                pivot = v

    for v in remaining_candidates:
        neighbours = graph.adjacencies[v]
        assert neighbours
        if pivot not in neighbours:
            candidates.remove(v)
            if neighbouring_candidates := candidates.intersection(neighbours):
                neighbouring_excluded = excluded.intersection(neighbours)
                visit(
                    graph=graph,
                    consumer=consumer,
                    pivot_choice_X=pivot_choice_X,
                    candidates=neighbouring_candidates,
                    excluded=neighbouring_excluded,
                    clique=clique + [v],
                )
            elif excluded.isdisjoint(neighbours):
                consumer.accept(clique + [v])
            excluded.add(v)

File: go/lib/bron_kerbosch_pivot.go
package BronKerbosch

type PivotSelection int

const (
	MaxDegree       PivotSelection = iota
	MaxDegreeLocal                 = iota
	MaxDegreeLocalX                = iota
)

func visit(graph *UndirectedGraph, reporter Reporter,
	initialPivotSelection PivotSelection, furtherPivotSelection PivotSelection,
	candidates VertexSet, excluded VertexSet, clique []Vertex) {
	if len(candidates) == 1 {
		for v := range candidates {
			// Same logic as below, stripped down
			neighbours := graph.neighbours(v)
			if excluded.IsDisjoint(neighbours) {
				reporter.Record(append(clique, v))
			}
		}
		return
	}

	var pivot Vertex
	remainingCandidates := make([]Vertex, 0, len(candidates))
	switch initialPivotSelection {
	case MaxDegree:
		pivot = pickMaxDegree(graph, candidates)
		for v := range candidates {
			remainingCandidates = append(remainingCandidates, v)
		}
	case MaxDegreeLocal, MaxDegreeLocalX:
		// Quickly handle locally unconnected candidates while finding pivot
		seenLocalDegree := 0
		for v := range candidates {
			neighbours := graph.neighbours(v)
			localDegree := neighbours.IntersectionLen(candidates)
			if localDegree == 0 {
				// Same logic as below, stripped down
				if neighbours.IsDisjoint(excluded) {
					reporter.Record(append(clique, v))
				}
			} else {
				if seenLocalDegree < localDegree {
					seenLocalDegree = localDegree
					pivot = v
				}
				remainingCandidates = append(remainingCandidates, v)
			}
		}
		if seenLocalDegree == 0 {
			return
		}
		if initialPivotSelection == MaxDegreeLocalX {
			for v := range excluded {
				neighbours := graph.neighbours(v)
				localDegree := neighbours.IntersectionLen(candidates)
				if seenLocalDegree < localDegree {
					seenLocalDegree = localDegree
					pivot = v
				}
			}
		}
	}

	for _, v := range remainingCandidates {
		neighbours := graph.neighbours(v)
		if neighbours.Contains(pivot) {
			continue
		}
		candidates.Remove(v)
		neighbouringCandidates := neighbours.Intersection(candidates)
		if !neighbouringCandidates.IsEmpty() {
			neighbouringExcluded := neighbours.Intersection(excluded)
			visit(
				graph, reporter,
				furtherPivotSelection, furtherPivotSelection,
				neighbouringCandidates,
				neighbouringExcluded,
				append(clique, v))
		} else {
			if neighbours.IsDisjoint(excluded) {
				reporter.Record(append(clique, v))
			}
		}
		excluded.Add(v)
	}
}

func pickMaxDegree(graph *UndirectedGraph, candidates VertexSet) Vertex {
	maxDegree := 0
	var maxVertex Vertex
	for v := range candidates {
		degree := graph.degree(v)
		if maxDegree < degree {
			maxDegree = degree
			maxVertex = v
		}
	}
	return maxVertex
}
